src/v/serde/rw/variant.h

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "base/vlog.h"
+#include "serde/rw/rw.h"
+
+#include <array>
+#include <type_traits>
+#include <utility>
+#include <variant>
+
+namespace serde {
+
+// A small wrapper around std::variant that is marked to be serializable.
+//
+// Special precation needs to be taken to mark a variant type as serializable
+// with respect to compatibility. `serde::variant` should be a drop in
+// replacement for `std::variant`, but allows for serde operations.
+//
+// # Variant Wire Compatibility:
+//
+// Variants are tricky w.r.t. compatibility, they are forward compatible if new
+// discrimiants are appended to the variant.
+//
+// Concretely that means that you can only read old versions of a variant if new
+// options are appended. Writing a new variant option should happen with a
+// feature barrier, so that an older reader will never read a newly written
+// variant.
+template<typename... Types>
+struct variant : public std::variant<Types...> {
+    using type = std::variant<Types...>;
+
+    constexpr variant() noexcept(std::is_nothrow_default_constructible_v<
+                                 std::variant_alternative_t<0, type>>)
+      = default;
+    constexpr variant(const variant&) noexcept(
+      std::is_nothrow_copy_constructible_v<type>)
+      = default;
+    constexpr variant(variant&&) noexcept(
+      std::is_nothrow_move_constructible_v<type>)
+      = default;
+
+    // Ensure that this is not implicitly convertable from std::variant
+    // but allow assignment from each individual type. For example:
+    //
+    // ```cpp
+    // using my_variant = serde::variant<int, bool>
+    //
+    // my_variant v = false; // should compile
+    //
+    // my_variant v = std::variant<int, bool>(false); // should NOT compile
+    // ```
+    template<class T>
+    constexpr variant(T&& t) noexcept // NOLINT(*-explicit-*)
```

Comment:
Missed this one

Reply:
Thanks! This is getting really hard to read all these annotations 😅